### PR TITLE
Implemented IIR filter for nAudio RSP HLE

### PR DESCRIPTION
**Pull request by kode54**
Specifically for CBFD

### CODE REVIEW

File: src/alist.c [hunk 1]
```diff
@@ -962,3 +962,64 @@ void alist_polef(
...
+        bool init,
+        uint16_t dmemo,
+        uint16_t dmemi,
+        uint16_t count,
+        int16_t* table,
+        uint32_t address)
+{
+    int16_t *dst = (int16_t*)(hle->alist_buffer + dmemo);
+    int32_t i, prev;
+	int16_t frame[8];
```
Comment (bsmiles32) [inline src/alist.c]: indentation

File: src/alist.c [hunk 2]
```diff
@@ -962,3 +962,64 @@ void alist_polef(
...
+    int16_t ibuf[4];
+    uint16_t index = 7;
+
+
+    count = align(count, 16);
+
+    #define vmulf(a, b)     (((a)*(b)+0x4000)>>15)
+
```
Comment (bsmiles32) [inline src/alist.c]: might be worth putting it inside arithmetic.h and as a static inline ?

File: src/alist.c [hunk 3]
```diff
@@ -962,3 +962,64 @@ void alist_polef(
...
+        ibuf[2] = (int16_t)*dram_u16(hle, address + 10);
+    }
+
+    prev = vmulf(table[9], frame[6]) * 2;
+    do
+    {
+        for(i = 0; i < 8; ++i)
+        {
+            int32_t accu;
+			ibuf[index&3] = *alist_s16(hle, dmemi);
```
Comment (bsmiles32) [inline src/alist.c]: indentation

**Comment by bsmiles32** [comment]
Thanks for your contribution !
The code looks good and is coherent with the rest of the codebase.
I've annotated small fixes.

I don't remember the details on what is expected of this audio command (I vaguely remember it to be a biquad filter of some sort), and I lost all my data recently, so I cannot quickly access that info anymore.
Do you confirm that this is exactly what the RSP does, or have you made some assumptions here ?

I'll try to test and review more carefully this week-end, but otherwise it looks good to me.

File: src/alist.c [hunk 4]
```diff
@@ -962,3 +962,64 @@ void alist_polef(
...
+void alist_iirf(
+        struct hle_t* hle,
+        bool init,
+        uint16_t dmemo,
+        uint16_t dmemi,
+        uint16_t count,
+        int16_t* table,
+        uint32_t address)
+{
+    int16_t *dst = (int16_t*)(hle->alist_buffer + dmemo);
```
Comment (bsmiles32) [inline src/alist.c]: Use alist_s16(hle, dmemo) instead

File: src/alist.c [hunk 5]
```diff
@@ -962,3 +962,64 @@ void alist_polef(
...
+            index=(index+1)&7;
+            dmemi += 2;
+        }
+        dst += 8;
+        count -= 0x10;
+    } while (count > 0);
+
+    dram_store_u16(hle, &frame[6], address + 4, 4);
+    dram_store_u16(hle, (int16_t*)&ibuf[(index-2)&3], address+8, 2);
+    dram_store_u16(hle, (int16_t*)&ibuf[(index-1)&3], address+10, 2);
```
Comment (bsmiles32) [inline src/alist.c]: for the 3 dram_store_u16 you should cast to (uint16_t*) the second parameter to avoid signedness warnings in gcc.

**Comment by bsmiles32** [comment]
@kode54 : Thanks for updating your PR :)
:+1:

**Comment by derselbst** [comment]
Sry for bringing this up again, but has there been any further research on what kind of filter this is? From a listen it sounds like a bandpass but looking at the code it looks more like lowpass.

**Comment by mudlord** [comment]
Resonant lowpass filter.

**Comment by derselbst** [comment]
Thanks!

**Pull request merged**